platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,1017 @@
+      Executor executor) {
+    Validate.notNull(core, "invalid null Solr core name");
+    Validate.notNull(creator, "invalid null Solr creator");
+    LOGGER.debug("Solr({}): Creating a Solr client adapter with creator [{}]", core, creator);
+    this.core = core;
+    this.creator = creator;
+    this.createFailsafe =
+        createFailsafeCreator
+            .apply(
+                SolrClientAdapter
+                    .ABORT_WHEN_INTERRUPTED_AND_RETRY_UNTIL_NO_ERROR_AND_A_CLIENT_IS_CREATED)
+            .with(SolrClientAdapter.SCHEDULED_EXECUTOR)
+            .onRetry(this::logFailure)
+            .onAbort(this::logInterruptionAndRecreate)
+            .onFailure(this::logAndRecreateIfNotCancelled)
+            .onSuccess(this::logAndSetConnecting);
+    this.pingFailsafe =
+        pingFailsafeCreator
+            .apply(SolrClientAdapter.ABORT_WHEN_INTERRUPTED_AND_RETRY_UNTIL_NO_ERROR)
+            .with(SolrClientAdapter.SCHEDULED_EXECUTOR)
+            .onRetry(this::logFailure)
+            .onAbort(this::logInterruptionAndReconnectIfStillConnecting)
+            .onFailure(this::logAndReconnectIfNotCancelledAndStillConnecting)
+            .onSuccess(this::logAndSetConnected);
+    this.waiter = waiter;
+    this.executor = executor;
+    this.lastPing = new AtomicLong(System.currentTimeMillis());
+    this.lastCreateError =
+        new AtomicLong(); // set to beginning of time to make sure we log the first error
+    this.unavailableClient =
+        new UnavailableSolrClient(
+            new UnavailableSolrException("initializing '" + core + "' client"));
+    this.apiClient = unavailableClient;
+    this.pingClient = unavailableClient;
+    this.realClient = null;
+    this.state = State.CREATING;
+    this.future = null;
+    setCreating(unavailableClient, false);
+  }
+
+  @Override
+  protected SolrClient getProxiedClient() {
+    if ((state != State.CONNECTED) && (state != State.CLOSED)) {
+      // do a spot check to see if it suddenly became reachable
+      try {
+        checkIfReachable("from the API because it is currently unavailable");
+        // if we get here than the ping was successful so make sure we move to connected
+        setConnected(true);
+        // fall-through to return the current one which should have been changed to the actual one
+      } catch (UnavailableSolrException e) {
+        // fall-through to return the current one which should still be an unavailable one
+      }
+    }
+    return this.apiClient;
+  }
+
+  @Override
+  protected <T> T handle(Code<T> code) throws SolrServerException, IOException {
+    try {
+      return code.invoke(getProxiedClient());
+    } catch (UnavailableSolrException e) {
+      throw e;
+    } catch (SolrException e) {
+      LOGGER.debug(
+          "Solr({}): API failure with code [{}] and metadata [{}]; {}",
+          core,
+          e.code(),
+          e.getMetadata(),
+          e, // this will get the info logged on the first line
```

Comment:
❓ Wouldn't `e.getMessage()` be clearer here than relying on the implicit `toString()`?

Reply:
`e.getMessage()` will only give me the message when sometimes the exception class also provides valuable info.